smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -0,0 +1,687 @@
+    for run_instance in persisted_runs:
+        run_entities: t.Dict[str, t.List[JobEntity]] = {
+            "model": [],
+            "orchestrator": [],
+            "ensemble": [],
+        }
+
+        for key in ["model", "orchestrator", "ensemble"]:
+            _entities = hydrate_persistables(key, run_instance, exp_dir)
+            for entity_type, new_entities in _entities.items():
+                if new_entities:
+                    run_entities[entity_type].extend(new_entities)
+
+        run = Run(
+            run_instance["timestamp"],
+            run_entities["model"],
+            run_entities["orchestrator"],
+            run_entities["ensemble"],
+        )
+        the_runs.append(run)
+
+    return the_runs
+
+
+def load_manifest(file_path: str) -> t.Optional[RuntimeManifest]:
+    """Load a persisted manifest and return the content"""
+    manifest_dict: t.Optional[t.Dict[str, t.Any]] = None
+    try_count = 1
+
+    while manifest_dict is None and try_count < 6:
+        source = pathlib.Path(file_path)
+        source = source.resolve()
+
+        try:
+            if text := source.read_text(encoding="utf-8").strip():
+                manifest_dict = json.loads(text)
+        except json.JSONDecodeError as ex:
+            print(f"Error loading manifest: {ex}")
+            # hack/fix: handle issues reading file before it is fully written
+            time.sleep(0.5 * try_count)
+        finally:
+            try_count += 1
+
+    if not manifest_dict:
+        return None
+
+    exp = manifest_dict.get("experiment", None)
+    if not exp:
+        raise ValueError("Manifest missing required experiment")
+
+    runs = manifest_dict.get("runs", None)
+    if runs is None:
+        raise ValueError("Manifest missing required runs")
+
+    exp_dir = pathlib.Path(exp["path"])
+    runs = hydrate_runs(runs, exp_dir)
+
+    manifest = RuntimeManifest(
+        name=exp["name"],
+        path=exp_dir,
+        launcher=exp["launcher"],
+        runs=runs,
+    )
+    return manifest
+
+
+def track_event(
+    timestamp: int,
+    task_id: t.Union[int, str],
+    step_id: str,
+    etype: str,
+    action: _EventClass,
+    status_dir: pathlib.Path,
+    logger: logging.Logger,
+    detail: str = "",
+    return_code: t.Optional[int] = None,
+) -> None:
+    """
+    Persist a tracking event for an entity
+    """
+    tgt_path = status_dir / f"{action}.json"
+    tgt_path.parent.mkdir(parents=True, exist_ok=True)
+
+    try:
+        task_id = int(task_id)
+    except ValueError:
+        pass
+
+    entity_dict = {
+        "timestamp": timestamp,
+        "job_id": task_id,
+        "step_id": step_id,
+        "type": etype,
+        "action": action,
+    }
+
+    if detail is not None:
+        entity_dict["detail"] = detail
+
+    if return_code is not None:
+        entity_dict["return_code"] = return_code
+
+    try:
+        if not tgt_path.exists():
+            """Don't overwrite existing tracking files"""
```

Comment:
One of my more pedantic requests:
```suggestion
            # Don't overwrite existing tracking files
```
I'm pretty sure that string gets optimized away at compile time, so it doesn't even save an alloc. Just felt better to have it as a comment hahahaha

Reply:
Agree 100. Maybe copilot created this monstrosity for me!? Yeah, blame the tool! Fixed